Remove race condition in actor framework

fulmicoton opened the pull request:
The bug could surfaced on any actor looping alone, like the control plane.

Closes #4248

fmassot reviewed (approved):
The last mailbox mechanism does not help us :/. I would open an issue to plan to drop it.

How did you test your fix?

fulmicoton commented:
@fmassot I agree! This is not the first time it causes us problems.

I switched back the PR to draft after I managed to reproduce despite the fix.
I found another bug, more straightforward. I added a unit test that one.
I'll keep testing and switch off the draft state.

fulmicoton commented on quickwit/quickwit-actors/src/mailbox.rs:
```diff
@@ -451,6 +419,7 @@ impl<A: Actor> WeakMailbox<A> {
     pub fn upgrade(&self) -> Option<Mailbox<A>> {
         let inner = self.inner.upgrade()?;
         let ref_count = self.ref_count.upgrade()?;
+        ref_count.fetch_add(1, Ordering::SeqCst);
```
that is the main bug here.

fulmicoton commented:
The bug seems fixed now. Tested with 
```
./target/release/quickwit-simian-army "chico(rate=1000000)*20" "chico(rate=100000)*20" "chico(rate=12000000)" "chico(rate=20000000)" --endpoint localhost:7280
```

over 15mn.
